cli/lint_reporter.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+//! This module defines an abstraction LintReporter, which is something that knows
+//! how to report observed lints (for example, pretty print or json).
+use crate::colors;
+use crate::fmt_errors;
+use crate::lint_serialize::SerializableLint;
+use deno_core::ErrBox;
+use deno_lint::diagnostic::LintDiagnostic;
+use serde::Serialize;
+
+pub enum LintReporterKind {
+  Console,
+
+  // `sort` is used to enforce determinism in tests.
+  Json { sort: bool },
+}
+
+pub trait LintReporter {
+  fn visit(&mut self, d: &LintDiagnostic);
+  fn visit_error(&mut self, file_path: &str, err: &ErrBox);
+  fn close(&mut self);
+}
+
+pub fn create_reporter(kind: LintReporterKind) -> Box<dyn LintReporter + Send> {
+  match kind {
+    LintReporterKind::Console => Box::new(ConsoleLintReporter::new()),
+    LintReporterKind::Json { sort } => Box::new(JsonLintReporter::new(sort)),
+  }
+}
+
+#[derive(Serialize)]
+struct LintError {
+  file_path: String,
+  message: String,
+}
+
+struct ConsoleLintReporter {
+  lint_count: u32,
+}
+
+impl ConsoleLintReporter {
+  fn new() -> ConsoleLintReporter {
+    ConsoleLintReporter { lint_count: 0 }
+  }
+}
+
+impl LintReporter for ConsoleLintReporter {
+  fn visit(&mut self, d: &LintDiagnostic) {
+    self.lint_count += 1;
+
+    let pretty_message =
+      format!("({}) {}", colors::gray(&d.code), d.message.clone());
+
+    let message = fmt_errors::format_stack(
+      true,
+      &pretty_message,
+      Some(&d.line_src),
+      Some(d.location.col as i64),
+      Some((d.location.col + d.snippet_length) as i64),
+      &[fmt_errors::format_location(
+        &d.location.filename,
+        d.location.line as i64,
+        d.location.col as i64,
+      )],
+      0,
+    );
+
+    eprintln!("{}\n", message);
+  }
+
+  fn visit_error(&mut self, file_path: &str, err: &ErrBox) {
+    eprintln!("Error linting: {}", file_path);
+    eprintln!("   {}", err);
+  }
+
+  fn close(&mut self) {
+    match self.lint_count {
+      1 => eprintln!("Found 1 problem"),
+      n if n > 1 => eprintln!("Found {} problems", self.lint_count),
+      _ => (),
+    }
+  }
+}
+
+#[derive(Serialize)]
+struct JsonLintReporter {
+  lints: Vec<SerializableLint>,
+  runtime_errors: Vec<LintError>,
```

Comment:
What is this field used for? Linter does only static analysis, so I don't expect to see any runtime errors

Reply:
> What is this field used for?

@bartlomieju this is a good question. Technically, [`lint_file`](https://github.com/denoland/deno/blob/fcea80391d014cf3ae3e15bdf3443fe4488443fb/cli/lint.rs#L65) returns a result object which could hold an `Err` signifying that something went wrong during the linting itself. Previously we'd just `eprintln` the error information (which is what the `PrettyLintReporter` does now), but if we did this with the JSON reporter, we'd break the JSON syntax of the CLI's output. So we needed someway to present this information in the JSON object itself.

I was thinking of calling this `errors`, but I thought that would be easily confused with "lint violations". What do you think? Is `errors` a better name here? Is there another name you think is more appropriate?

Reply:
I see, that makes sense. I think we should go with `diagnostics` and `errors`.

Reply:
> I see, that makes sense. I think we should go with diagnostics and errors.

Done! Thanks @bartlomieju!